---
layout: workshop
root: .
venue: Indonesia Biodiversity Research Center, Bali      # brief name of host site
address: Praja Hotel, Renon, Denpasar, Bali    # street address of home location
country: Indonesia    # country (hyphenated if need be, like 'United-Kingdom')
language: en   # language code like 'en' or 'fr'
latlng: -8.671308,115.239431     # fractional latitude and longitude like 41.7901128,-87.6007318
humandate: Aug 3-4, 2015  # use the format 'Feb 17-18, 2020' (without the quotes)
humantime: 9:00 am - 5:00 pm  # use the format '9:00 am - 4:30 pm' (without the quotes)
startdate: 2015-08-03  # use YYYY-MM-DD format like 2015-01-01
enddate: 2015-08-04    # use YYYY-MM-DD format like 2015-01-02
instructor: ["Areej Alsheikh-Hussain","Eric Crandall"] # list of names like ["Kay McNulty", "Betty Jennings", "Betty Snyder"]
helper: ["John Deck","Katja Lehmann","Cynthia Riginos","Patricia Cockett"]     # list of names like ["Marlyn Wescoff", "Fran Bilas", "Ruth Lichterman"]
contact: eric.d.crandall@gmail.com    # contact email address for workshop organizer
etherpad: https://etherpad.mozilla.org/2015-08-03-IBRC-Indonesia  # optional (insert the URL for your Etherpad if you're using one)
---
<!--
  HEADER

  Edit the values in the block above to be appropriate for your
  workshop.  Run 'tools/check' *before* committing to make sure that
  changes are good.
-->

<!--
  EVENTBRITE

  This block includes the Eventbrite registration widget if
  'eventbrite' has been set in the header.  You can delete it if you
  are not using Eventbrite, or leave it in, since it will not be
  displayed if the 'eventbrite' field in the header is not set.
-->
{% if page.eventbrite %}
<iframe
  src="https://www.eventbrite.com/tickets-external?eid={{page.eventbrite}}&ref=etckt"
  frameborder="0"
  width="100%"
  height="206px"
  scrolling="auto">
</iframe>
{% endif %}

<h2>General Information</h2>

<!--
  INTRODUCTION

  Edit the general explanatory paragraph below if you want to change
  the pitch.
-->
<p>
  These lessons are intended to give learners a general introduction to computational skills that are increasingly important for biologists to understand.
  Do not expect to emerge from this workshop as an expert in computational biology! 
  Our overall goal is to give you a foundation of skills that you can extend through using them in your research.

</p> 
<p>
  Software Carpentry's mission
  is to help scientists and engineers get more research done in less
  time and with less pain by teaching them basic lab skills for
  scientific computing.  This hands-on workshop will cover basic
  concepts and tools, including program design, data
  management, and task automation.  Participants will be encouraged to
  help one another and to apply what they have learned to their own
  research problems.
</p>
<p align="center">
  <em>
    For more information on what we teach and why,
    please see our paper
    "<a href="http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1001745">Best Practices for Scientific Computing</a>".
  </em>
</p>

<!--
  AUDIENCE

  Explain who your audience is.  (In particular, tell readers if the
  workshop is only open to people from a particular institution.
-->
<p>
  <strong>Who:</strong>
  The course is aimed at biology faculty from Southeast Asia
</p>

<!--
  LOCATION

  This block displays the address and links to maps showing directions
  if the latitude and longitude of the workshop have been set.  You
  can use http://itouchmap.com/latlong.html to find the lat/long of an
  address.
-->
{% if page.latlng %}
<p>
  <strong>Where:</strong>
  {{page.address}}.
  Get directions with
  <a href="//www.openstreetmap.org/?mlat={{page.latlng | replace:',','&mlon='}}&zoom=16">OpenStreetMap</a>
  or
  <a href="//maps.google.com/maps?q={{page.latlng}}">Google Maps</a>.
</p>
{% endif %}

<!--
  SPECIAL REQUIREMENTS

  Modify the block below if there are any special requirements.
-->
<p>
  <strong>Requirements:</strong> Participants must bring a laptop with
  a few specific software packages installed (listed
  <a href="#setup">below</a>). They are also required to abide by
  Software Carpentry's
  <a href="{{site.swc_site}}/conduct.html">Code of Conduct</a>.
</p>

<!--
  CONTACT EMAIL ADDRESS

  Display the contact email address set in the header.  If an address
  isn't set in the header, the Software Carpentry admin address is
  used.
-->
<p>
  <strong>Contact</strong>:
  Please mail
  {% if page.contact %}
    <a href='mailto:{{page.contact}}'>{{page.contact}}</a>
  {% else %}
    <a href='mailto:{{site.contact}}'>{{site.contact}}</a>
  {% endif %}
  for more information.
</p>

<hr/>

<!--
  SCHEDULE

  Show the workshop's schedule.  Edit the items and times in the table
  to match your plans.  You may also want to change 'Day 1' and 'Day
  2' to be actual dates or days of the week.
-->
<h2>Schedule</h2>
<p> Note that this schedule only applies to the Software Carpentry portion of the workshop (first two days)</p>
<div class="row">
  <div class="col-md-6">
    <h3>Day 1</h3>
    <table class="table table-striped">
      <tr> <td>10:00</td>  <td>Introduction, software setup and troubleshooting</td> </tr>
      <tr> <td>11:00</td>  <td>Automating tasks with the Unix shell</td> </tr>
      <tr> <td>12:00</td>  <td>Lunch</td> </tr>
      <tr> <td>13:30</td>  <td>Automating tasks with the Unix shell - continued</td> </tr>
      <tr> <td>15:00</td>  <td>Break</td> </tr>
      <tr> <td>15:30</td>  <td>Automating tasks with the Unix shell - continued</td> </tr>
      <tr> <td>17:00</td>  <td>Wrap-up</td> </tr>
    </table>
  </div>
  <div class="col-md-6">
    <h3>Day 2</h3>
    <table class="table table-striped">
      <tr> <td>9:00</td>  <td>Programming with R</td> </tr>
      <tr> <td>10:30</td>  <td>Break</td> </tr>
       <tr> <td>10:45</td>  <td>Programming with R - continued</td> </tr>
      <tr> <td>12:00</td>  <td>Lunch</td> </tr>
      <tr> <td>13:30</td>  <td>Programming with R - continued</td> </tr>
      <tr> <td>15:00</td>  <td>Break</td> </tr>
      <tr> <td>15:30</td>  <td>Programming with R - continued</td> </tr>
      <tr> <td>16:30</td>  <td>Wrap-up for Software Carpentry Workshop</td> </tr>
    </table>
  </div>
</div>

<!--
  ETHERPAD

  Display the Etherpad for the workshop.  You can set this up in
  advance or on the first day; either way, make sure you push changes
  to GitHub after you have its URL.  To create an Etherpad, go to

      http://etherpad.mozilla.org/YYYY-MM-DD-site

  where 'YYYY-MM-DD-site' is the identifier for your workshop (e.g.,
  '2015-06-10-esu'.
-->
{% if page.etherpad %}
<p>
  <strong>Etherpad:</strong> <a href="{{page.etherpad}}">{{page.etherpad}}</a>.
  <br/>
  We will use this Etherpad for chatting, taking notes, and sharing URLs and bits of code.
</p>
{% endif %}

<hr/>

<!--
  SYLLABUS

  Show what topics will be covered.

  1. If your workshop is R rather than Python, remove the comment
     around that section and put a comment around the Python section.
  2. Some workshops will delete SQL.
  3. Please make sure the list of topics is synchronized with what you
     intend to teach.
  4. You may need to move the div's with class="col-md-6" around inside
     the div's with class="row" to balance the multi-column layout.

  This is one of the places where people frequently make mistakes, so
  please preview your site before committing, and make sure to run
  'tools/check' as well.
-->
<h2>Syllabus</h2>

<div class="row">
  <div class="col-md-6">
    <h3>The Unix Shell</h3>
    <ul>
      <li>Files and directories</li>
      <li>History and tab completion</li>
      <li>Pipes and redirection</li>
      <li>Looping over files</li>
      <li>Creating and running shell scripts</li>
      <li>Finding things</li>
      <li><a href="{{site.swc_lessons}}/ref/01-shell.html">Reference...</a></li>
      <li><a href="http://swcarpentry.github.io/shell-novice/">Lessons...</a></li>
    </ul>
  </div>
  <!--
  <div class="col-md-6">
    <h3>Programming in Python</h3>
    <ul>
      <li>Using libraries</li>
      <li>Working with arrays</li>
      <li>Reading and plotting data</li>
      <li>Creating and using functions</li>
      <li>Loops and conditionals</li>
      <li>Defensive programming</li>
      <li>Using Python from the command line</li>
      <li><a href="{{site.swc_lessons}}/ref/03-python.html">Reference...</a></li>
    </ul>
  </div>
  -->
  
  <div class="col-md-6">
    <h3>Programming in R</h3>
    <ul>
      <li>Working with vectors and data frames</li>
      <li>Reading and plotting data</li>
      <li>Creating and using functions</li>
      <li>Loops and conditionals</li>
      <li>Using R from the command line</li>
      <li><a href="{{site.swc_lessons}}/ref/06-R.html">Reference...</a></li>
      <li><a href="http://swcarpentry.github.io/r-novice-inflammation/">Lessons...</a></li>
    </ul>
  </div>

</div>

<div class="row">
<!--
  <div class="col-md-6">
    <h3>Version Control with Git</h3>
    <ul>
      <li>Creating a repository</li>
      <li>Recording changes to files: <code>add</code>, <code>commit</code>, ...</li>
      <li>Viewing changes: <code>status</code>, <code>diff</code>, ...</li>
      <li>Ignoring files</li>
      <li>Working on the web: <code>clone</code>, <code>pull</code>, <code>push</code>, ...</li>
      <li>Resolving conflicts</li>
      <li>Open licenses</li>
      <li>Where to host work, and why</li>
      <li><a href="{{site.swc_lessons}}/ref/02-git.html">Reference...</a></li>
    </ul>
  </div>
  <div class="col-md-6">
    <h3>Managing Data with SQL</h3>
    <ul>
      <li>Reading and sorting data</li>
      <li>Filtering with <code>where</code></li>
      <li>Calculating new values on the fly</li>
      <li>Handling missing values</li>
      <li>Combining values using aggregation</li>
      <li>Combining information from multiple tables using <code>join</code></li>
      <li>Creating, modifying, and deleting data</li>
      <li>Programming with databases</li>
      <li><a href="{{site.swc_lessons}}/ref/04-sql.html">Reference...</a></li>
    </ul>
  </div>
</div>
-->
<hr/>
<!--
  SETUP

  Delete irrelevant sections from the setup instructions.  Each
  section is inside a 'div' without any classes to make the beginning
  and end easier to find.

  This is the other place where people frequently make mistakes, so
  please preview your site before committing, and make sure to run
  'tools/check' as well.
-->
<h2 id="setup">Setup</h2>

<p>
  To participate in a Software Carpentry workshop, you will need
  access to the software described below. In addition, you will
  need an up-to-date web browser.
  We maintain a list of common issues that occur during installation as a reference for instructors
  that may be useful on the
  <a href = "https://github.com/swcarpentry/workshop-template/wiki/Configuration-Problems-and-Solutions">Configuration Problems and Solutions wiki page</a>.
</p>

<div id="shell"> <!-- Start of 'shell' section. -->
  <h3>The Bash Shell</h3>

  <p>
    Bash is a commonly-used shell that gives you the power to do simple
    tasks more quickly. We will be using 
    <a href= "http://environmentalomics.org/bio-linux/">Bio-Linux</a> as a common platform for learning. Bio-Linux is a distribution of the popular and free Ubuntu Linux
    operating system, but includes more than 250 useful bioinformatic programs and packages. We will be using Bio-Linux for the Introduction to the Unix Shell, but also for lessons on RAD and Metagenomic pipelines on Thursday and Friday.
    You may install and run Bio-Linux on your computer in a variety of ways:
  </p>

  <div class="row">
    <div class="col-md-4">
      <h4 id="shell-windows">Windows</h4>
      <p>
        We will provide Bio-Linux on USB drives (generously provided by Katja Lehmann and the UK National Environment Research Council (NERC).
        You may boot your Windows computer directly from these drives to run Bio-Linux. Please refer <a href= "http://environmentalomics.org/bio-linux-installation/">here</a> for instructions ahead of time. 
        <strong>You will need to figure out how to boot your computer from a USB drive ahead of time. Please do so before class starts on Monday</strong></p>
      <p>
        Alternatively, you may install <a href="http://virtualbox.org">VirtualBox</a> (free emulation software) and install Bio-Linux on this. This will allow you to run Windows and Bio-Linux simultaneously.
        See instructions <a href= "http://environmentalomics.org/bio-linux-installation/">here</a> (Running Bio-Linux as a VM with VirtualBox) as well as below. This option requires 40GB of free disk space, and a 4GB download.
        Eric Crandall will also have the 4GB .ova file available on Sunday and Monday morning for those who are interested in this option.</p>
      <p>
        If neither of the above options work, please consider installing Git for Windows by downloading and running
        <a href="http://msysgit.github.io/">the installer</a>.
        This will provide you with both Git and Bash in the Git Bash program. However, you will need to install a number of additional programs if you want to use your computer for lessons on Metagenomics and RAD pipelines.</p>
       <p>
        Finally, there is the possibility of using the <a href="https://resbaz.cloud.edu.au"> ResBaz</a> platform, which will allow you to use R and Unix from your browser.
        Please talk to an instructor if you feel this option is for you (you are unable to install anything). If you use this option, you need to have a Twitter or Github account beforehand.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="shell-macosx">Mac OS X</h4>
      <p>
        The USB drives cannot be booted from a Mac. We therefore recommend that you install <a href="http://virtualbox.org">VirtualBox</a> (free emulation software) and install Bio-Linux on this. This will allow you to run Windows and Bio-Linux simultaneously.
        See instructions <a href= "http://environmentalomics.org/bio-linux-installation/">here</a> (Running Bio-Linux as a VM with VirtualBox) as well as below. This option requires 40GB of free disk space, and a 4GB download.
        Eric Crandall will have the .ova file available on Sunday and Monday morning for those who are interested in this option.</p>
       <p> 
        Fortunately, the default shell in all versions of Mac OS X is bash, so if you are not able to install Bio-Linux you can use this with no
        need to install anything. However, you will need to install a number of additional programs if you want to use your computer for lessons on Metagenomics and RAD pipelines. 
        Please consult the instructors about this. 
        You access bash from the Terminal
        (found in
        <code>/Applications/Utilities</code>).  You may want to keep
        Terminal in your dock for this workshop.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="shell-linux">Linux</h4>
      <p>
        The default shell is usually Bash, but if your
        machine is set up differently you can run it by opening a
        terminal and typing <code>bash</code>.  You will need to install a number of additional programs if you want to use your computer  for lessons on Metagenomics and RAD pipelines.  
      </p>
    </div>
  </div>
</div> <!-- End of 'shell' section. -->

<div id="editor"> <!-- Start of 'editor' section. -->
  <h3>Text Editor</h3>

  <p>
    When you're writing code, it's nice to have a text editor that is
    optimized for writing code, with features like automatic
    color-coding of key words.  The default text editor on Mac OS X and
    Linux is usually set to Vim, which is not famous for being
    intuitive.  if you accidentally find yourself stuck in it, try
    typing the escape key, followed by <code>:q!</code> (colon, lower-case 'q',
    exclamation mark), then hitting Return to return to the shell.
  </p>

  <div class="row">
    <div class="col-md-4">
      <h4 id="editor-windows">Windows</h4>
      
      <p>
        Vim and Emacs are two basic editors that come in bash. They are a bit tricky to use.
        If you are booted into Bio-Linux, then use <a href="https://wiki.gnome.org/Apps/Gedit">Gedit</a>, the default windowed text editor, which is fairly user friendly.
        There are a number of Windows user-friendly editors that you can use, but you won't be able to use them if you are booted using a USB drive:
        <a href="http://notepad-plus-plus.org/">Notepad++</a> or
        <a href="http://www.sublimetext.com/">Sublime Text</a>.
        <strong>Be aware that you must
          add its installation directory to your system path.</strong>
        Please ask your instructor to help you do this.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="editor-macosx">Mac OS X</h4>
      <p>
        If you are booted into Bio-Linux, then use <a href="https://wiki.gnome.org/Apps/Gedit">Gedit</a>, the default windowed text editor, which is fairly user friendly.
        Besides Vim and Emacs, other text editors that you can use are:
        <a href="http://www.barebones.com/products/textwrangler/">Text Wrangler</a> or
        <a href="http://www.sublimetext.com/">Sublime Text</a>.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="editor-linux">Linux</h4>
      <p>
        Besides Vim and Emacs, other text editors that you can use are:
        <a href="https://wiki.gnome.org/Apps/Gedit">Gedit</a>,
        <a href="http://kate-editor.org/">Kate</a> or
        <a href="http://www.sublimetext.com/">Sublime Text</a>.
      </p>
    </div>
  </div>
</div> <!-- End of 'editor' section. -->


<div id="r"> <!-- Start of 'R' section. -->
  <h3>R</h3>

  <p>
    <a href="http://www.r-project.org">R</a> is a programming language
    that is especially powerful for data exploration, visualization, and
    statistical analysis. To interact with R, we use
    <a href="http://www.rstudio.com/">RStudio</a>. </p>
   <p> If R installation doesn't work for you, 
    there is the possibility of using the <a href="https://resbaz.cloud.edu.au"> ResBaz</a> platform, 
    which will allow you to use R and Rstudio from your browser. 
    Please talk to an instructor if you feel this option is for you. If you use this option, you need to have a Twitter or Github account beforehand.
  </p>

  <div class="row">
    <div class="col-md-4">
      <h4 id="r-windows">Windows</h4>
      <p>
        Install R by downloading and running
        <a href="http://cran.r-project.org/bin/windows/base/release.htm">this .exe file</a>
        from <a href="http://cran.r-project.org/index.html">CRAN</a>.
        Also, please install the
        <a href="http://www.rstudio.com/ide/download/desktop">RStudio IDE</a>.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="r-macosx">Mac OS X</h4>
      <p>
        Install R by downloading and running
        <a href="http://cran.r-project.org/bin/macosx/R-latest.pkg">this .pkg file</a>
        from <a href="http://cran.r-project.org/index.html">CRAN</a>.
        Also, please install the
        <a href="http://www.rstudio.com/ide/download/desktop">RStudio IDE</a>.
      </p>
    </div>
    <div class="col-md-4">
      <h4 id="r-linux">Linux</h4>
      <p>
        You can download the binary files for your distribution
        from <a href="http://cran.r-project.org/index.html">CRAN</a>. Or
        you can use your package manager (e.g. for Debian/Ubuntu
        run <code>sudo apt-get install r-base</code> and for Fedora run
        <code>sudo yum install R</code>).  Also, please install the
        <a href="http://www.rstudio.com/ide/download/desktop">RStudio IDE</a>.
      </p>
    </div>
  </div>
</div> <!-- End of 'R' section. -->

<div id="vm">
  <h3>Virtual Machine</h3>

  <p>
    Here are further instructions for downloading and installing VirtualBox
  </p>
  <ol>
    <li>
      Install <a href="https://www.virtualbox.org/">VirtualBox</a>.
    </li>
    <li>
      Download the Bio-Linux <a href="http://nebc.nerc.ac.uk/downloads/bio-linux-8-latest.iso">VM image</a>.
      <strong>Warning:</strong> this file is >4 GByte, so please
      download it <em>before</em> coming to your workshop.
    </li>
    <li>
      Load the VM into VirtualBox by selecting "Import Appliance" and
      loading the <code>.ova</code> file.
    </li>
  </ol>
</div>

